src/main/java/spoon/pattern/PatternBuilder.java

```diff
@@ -0,0 +1,935 @@
+		modifyNodeOfAttributeOfElement(element, role, conflictMode, oldAttrNode -> {
+			return node;
+		});
+	}
+
+	/**
+	 * @param element to be checked element
+	 * @return true if element `element` is a template or a child of template
+	 */
+	public boolean isInModel(CtElement element) {
+		if (element != null) {
+			for (CtElement patternElement : patternModel) {
+				if (element == patternElement || element.hasParent(patternElement)) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	public Pattern build() {
+		if (built) {
+			throw new SpoonException("The Pattern may be built only once");
+		}
+		built = true;
+		//clean the mapping so it is not possible to further modify built pattern using this builder
+		patternElementToSubstRequests.clear();
+		return new Pattern(new ModelNode(patternNodes.getNodes())).setAddGeneratedBy(isAddGeneratedBy());
+	}
+
+	static List<? extends CtElement> bodyToStatements(CtStatement statementOrBlock) {
+		if (statementOrBlock instanceof CtBlock) {
+			return ((CtBlock<?>) statementOrBlock).getStatements();
+		}
+		return Collections.singletonList(statementOrBlock);
+	}
+
+	/**
+	 * @return default {@link ValueConvertor}, which will be assigned to all new {@link ParameterInfo}s
+	 */
+	public ValueConvertor getDefaultValueConvertor() {
+		return valueConvertor;
+	}
+
+	/**
+	 * @param valueConvertor default {@link ValueConvertor}, which will be assigned to all {@link ParameterInfo}s created after this call
+	 * @return this to support fluent API
+	 */
+	public PatternBuilder setDefaultValueConvertor(ValueConvertor valueConvertor) {
+		this.valueConvertor = valueConvertor;
+		return this;
+	}
+
+	/**
+	 * All the variable references, whose variables are out of the template model
+	 * are automatically marked as pattern parameters
+	 * @return this to support fluent API
+	 */
+	public PatternBuilder configureAutomaticParameters() {
+		configureParameters(pb -> {
+			//add this substitution request only if there isn't another one yet
+			pb.setConflictResolutionMode(ConflictResolutionMode.KEEP_OLD_NODE);
+			/*
+			 * detect other parameters.
+			 * contract: All variable references, which are declared outside of template are automatically considered as template parameters
+			 */
+			pb.queryModel().filterChildren(new TypeFilter<>(CtVariableReference.class))
+				.forEach((CtVariableReference<?> varRef) -> {
+					CtVariable<?> var = varRef.getDeclaration();
+					if (var == null || isInModel(var) == false) {
+						//the varRef has declaration out of the scope of the template. It must be a template parameter.
+						ParameterInfo parameter = pb.parameter(varRef.getSimpleName()).getCurrentParameter();
+						pb.addSubstitutionRequest(parameter, varRef);
+					}
+				});
+		});
+		return this;
+	}
+
+	/**
+	 * Builder which allows to define which part of AST of template {@link CtType}
+	 * has to be used as a model of the {@link Pattern}
+	 */
+	public static class TemplateModelBuilder {
```

Comment:
Why do you keep this class internal? `PatternBuilder` is already pretty big, it seems easier to maintain/get access to the doc if the class is in its own CompilationUnit.

Reply:
good idea, I will move it to own file